scipy/stats/_qmc_cy.pyx

```diff
@@ -0,0 +1,326 @@
+cdef double l2_star_discrepancy(double[:, ::1] sample_view,
+                             bint iterative, unsigned int workers) nogil:
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double prod = 1, disc1 = 0
+
+    for i in range(n):
+        for j in range(d):
+                prod *= 1 - sample_view[i, j] ** 2
+
+        disc1 += prod
+        prod = 1
+
+    cdef double disc2 = threaded_loops(l2_star_loop, sample_view, workers)
+
+    if iterative:
+        n += 1
+
+    cdef double one_div_n = <double> 1 / n
+    return sqrt(
+        pow(3, -d) - one_div_n * pow(2, 1 - d) * disc1 + 1 / pow(n, 2) * disc2
+    )
+
+
+cdef double l2_star_loop(double[:, ::1] sample_view, Py_ssize_t istart,
+                         Py_ssize_t istop) nogil:
+
+    cdef:
+        Py_ssize_t i, j, k
+        double prod = 1, disc2 = 0, tmp_sum = 0
+
+    for i in range(istart, istop):
+        for j in range(sample_view.shape[0]):
+            prod = 1
+            for k in range(sample_view.shape[1]):
+                prod *= (
+                        1 - max(sample_view[i, k], sample_view[j, k])
+                )
+            tmp_sum += prod
+
+        disc2 += tmp_sum
+        tmp_sum = 0
+
+    return disc2
+
+
+def _cy_wrapper_update_discrepancy(double[::1] x_new_view,
+                                 double[:, ::1] sample_view,
+                                 double initial_disc):
+    return c_update_discrepancy(x_new_view, sample_view, initial_disc)
+
+
+cdef double c_update_discrepancy(double[::1] x_new_view,
+                                 double[:, ::1] sample_view,
+                                 double initial_disc):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0] + 1
+        Py_ssize_t xnew_nlines = x_new_view.shape[0]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double prod = 1, tmp_sum= 0
+        double  disc1 = 0, disc2 = 0, disc3 = 0
+        double[::1] abs_ = np.zeros(n, dtype=np.float64)
+
+
+    # derivation from P.T. Roy (@tupui)
+    for i in range(xnew_nlines):
+        abs_[i] = fabs(x_new_view[i] - 0.5)
+        prod *= (
+                    1 + 0.5 * abs_[i]
+                    - 0.5 * pow(abs_[i], 2)
+        )
+
+    disc1 = (- 2 / <double> n) * prod
+
+    prod = 1
+    for i in range(n - 1):
+        for j in range(xnew_nlines):
+            prod *= (
+                1 + 0.5 * abs_[j] +
+                0.5 * fabs(sample_view[i, j] - 0.5) -
+                0.5 * fabs(x_new_view[j] - sample_view[i, j])
+            )
+        disc2 += prod
+        prod = 1
+
+    disc2 *= 2 / pow(n, 2)
+
+    for i in range(xnew_nlines):
+        prod *= 1 + abs_[i]
+
+    disc3 = 1 / pow(n, 2) * prod
+
+    return initial_disc + disc1 + disc2 + disc3
+
+
+ctypedef double (*func_type)(double[:, ::1], Py_ssize_t,
+                             Py_ssize_t) nogil
+
+
+cdef double threaded_loops(func_type loop_func, double[:,
```

Comment:
The threading part might be something we might want to have in general. @rgommers maybe we could share c++/cython utils somewhere?

Reply:
This part is a little bit specific. The data must be a vlew, the function must work on a reduction with size the number of rows of the view. In my pov this is too specific to be in a more general util. Maybe this code could be adapted to be more general, but this can cause a overhead, which can be problematic, and this can lead to a understandably problem.

Reply:
ah ok thanks for the clarification. I hope this was general enough so that we could easily give an "official way" to have `workers` in cython.